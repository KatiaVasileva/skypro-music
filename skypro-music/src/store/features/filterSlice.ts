import { createSlice, PayloadAction } from "@reduxjs/toolkit";

type FilterStateType = {
  searchState: string;
};

const initialState: FilterStateType = {
  searchState: "",
};

const filterSlice = createSlice({
  name: "filter",
  initialState,
  reducers: {
    setSearchState: (state, action: PayloadAction<string>) => {
      state.searchState = action.payload;
    },
  },
});

export const { setSearchState } = filterSlice.actions;
export const filterReducer = filterSlice.reducer;
