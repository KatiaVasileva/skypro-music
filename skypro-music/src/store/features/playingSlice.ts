import { createSlice, PayloadAction } from "@reduxjs/toolkit";

type PlayingStateType = {
  playingState: boolean;
};

const initialState: PlayingStateType = {
  playingState: false,
};

const playingSlice = createSlice({
  name: "playing",
  initialState,
  reducers: {
    setPlayingState: (state, action: PayloadAction<boolean>) => {
      state.playingState = action.payload;
    },
    togglePlaying: (state) => {
      state.playingState = !state.playingState;
    },
  },
});

export const { setPlayingState, togglePlaying } = playingSlice.actions;
export const playingReducer = playingSlice.reducer;
