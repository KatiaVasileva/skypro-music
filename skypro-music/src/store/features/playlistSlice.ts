import { Track } from "@/types/Track.types";
import { createSlice, PayloadAction } from "@reduxjs/toolkit";

type PlaylistStateType = {
    playlistState: Array<Track>;
};

const initialState: PlaylistStateType = {
    playlistState: [],
};

const playlistSlice = createSlice({
    name: "playlist",
    initialState,
    reducers: {
        setPlaylistState: (state, action: PayloadAction<Array<Track>>) => {
            state.playlistState = action.payload;
        },
    },
});

export const {setPlaylistState} = playlistSlice.actions;
export const playlistReducer = playlistSlice.reducer;