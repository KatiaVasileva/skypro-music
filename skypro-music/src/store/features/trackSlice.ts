import {
  addFavorite,
  FavoriteRequestProps,
  getAllTracks,
  getFavorite,
  removeFavorite,
} from "@/api/tracksApi";
import { Track } from "@/types/Track.types";
import { createAsyncThunk, createSlice, PayloadAction } from "@reduxjs/toolkit";

type TrackStateType = {
  trackState?: Track;
  trackIndexState: number;
  playingState: boolean;
  playlistState: Array<Track>;
  shuffledPlaylistState: Array<Track>;
  myPlaylistState: Array<Track>;
  shuffleActiveState: boolean;
  isLiked: boolean;
  isMyPlaylistClicked: boolean;
  errorMessage: string;
};

const initialState: TrackStateType = {
  trackState: undefined,
  trackIndexState: -1,
  playingState: false,
  playlistState: [],
  shuffledPlaylistState: [],
  myPlaylistState: [],
  shuffleActiveState: false,
  isLiked: false,
  isMyPlaylistClicked: false,
  errorMessage: "",
};

export const getTracks = createAsyncThunk("track/allTracks", async () => {
  const allTracks = await getAllTracks();
  return allTracks;
});

export const getFavoriteTracks = createAsyncThunk(
  "track/favorite",
  async ({ access, refresh }: { access: string; refresh: string }) => {
    const myTracks = await getFavorite({ access, refresh });
    return myTracks;
  }
);

export const addTrackInFavorite = createAsyncThunk(
  "track/addFavorite",
  async ({ id, access, refresh }: FavoriteRequestProps) => {
    const likedTrack = await addFavorite({ id, access, refresh });
    return likedTrack;
  }
);

export const removeTrackFromFavorite = createAsyncThunk(
  "track/removeFavorite",
  async ({ id, access, refresh }: FavoriteRequestProps) => {
    const dislikedTrack = await removeFavorite({ id, access, refresh });
    return dislikedTrack;
  }
);

const trackSlice = createSlice({
  name: "track",
  initialState,
  reducers: {
    setTrackState: (state, action: PayloadAction<Track>) => {
      state.trackState = action.payload;
    },
    setPlaylistState: (
      state,
      action: PayloadAction<{ tracks: Array<Track> }>
    ) => {
      if (state.shuffleActiveState) {
        state.shuffledPlaylistState = [...action.payload.tracks].sort(
          () => 0.5 - Math.random()
        );
      } else {
        state.playlistState = action.payload.tracks;
      }
    },
    setMyPlaylistState: (state, action: PayloadAction<Array<Track>>) => {
      state.myPlaylistState = action.payload;
    },
    setTrackIndexState: (state, action: PayloadAction<number>) => {
      state.trackIndexState = action.payload;
    },
    setPlayingState: (state, action: PayloadAction<boolean>) => {
      state.playingState = action.payload;
    },
    setShuffleActiveState: (state, action: PayloadAction<boolean>) => {
      state.shuffleActiveState = action.payload;
    },
    togglePlaying: (state) => {
      state.playingState = !state.playingState;
    },
    toggleShuffle: (state) => {
      state.shuffleActiveState = !state.shuffleActiveState;
    },
    setNextTrack: (state) => {
      const playlist = state.shuffleActiveState
        ? state.shuffledPlaylistState
        : state.playlistState;
      const nextIndex =
        state.trackIndexState < playlist.length - 1
          ? state.trackIndexState + 1
          : 0;
      state.trackIndexState = nextIndex;
      state.trackState = playlist[nextIndex];
    },
    setPrevTrack: (state) => {
      const playlist = state.shuffleActiveState
        ? state.shuffledPlaylistState
        : state.playlistState;
      const nextIndex =
        state.trackIndexState > 0 ? state.trackIndexState - 1 : 0;
      state.trackIndexState = nextIndex;
      state.trackState = playlist[nextIndex];
    },
    toggleIsLiked: (state) => {
      if (state.trackState) {
        state.isLiked = !state.isLiked;
      }
    },
    setIsMyPlaylistClicked: (state, action: PayloadAction<boolean>) => {
      state.isMyPlaylistClicked = action.payload;
    },
    setLike: (state, action: PayloadAction<Track>) => {
      state.myPlaylistState.push(action.payload);
    },
    setDislike: (state, action: PayloadAction<Track>) => {
      state.myPlaylistState = state.myPlaylistState.filter(
        (track) => track._id !== action.payload._id
      );
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(getTracks.fulfilled, (state, action) => {
        state.playlistState = action.payload;
      })
      .addCase(getTracks.rejected, (state, action) => {
        state.errorMessage = "Ошибка: " + action.error.message;
      })
      .addCase(getFavoriteTracks.fulfilled, (state, action) => {
        state.myPlaylistState = action.payload;
      })
      .addCase(getFavoriteTracks.rejected, (state, action) => {
        state.errorMessage = "Ошибка: " + action.error.message;
      });
  },
});

export const {
  setTrackState,
  setTrackIndexState,
  setPlayingState,
  togglePlaying,
  setPlaylistState,
  setMyPlaylistState,
  setNextTrack,
  setPrevTrack,
  setShuffleActiveState,
  toggleShuffle,
  toggleIsLiked,
  setIsMyPlaylistClicked,
} = trackSlice.actions;
export const trackReducer = trackSlice.reducer;
