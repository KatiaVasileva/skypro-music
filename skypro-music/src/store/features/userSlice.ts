import { getToken, GetTokenProps, login, LoginProps, refreshToken, register, RegisterProps } from "@/api/userApi";
import { Tokens } from "@/types/Tokens.types";
import { User } from "@/types/User.types";
import {
  getUserFromLocalStorage,
  removeUserFromLocalStorage,
  saveUserToLocalStorage,
} from "@/utils/helpers";
import { createAsyncThunk, createSlice, PayloadAction } from "@reduxjs/toolkit";

type UserStateType = {
  userState: User | null;
  isAuthState: boolean;
  errorMessage: string;
  tokens: Tokens | null;
};

const userFromLocalStorage: User | null = getUserFromLocalStorage();

const initialState: UserStateType = {
  userState: userFromLocalStorage,
  isAuthState: false,
  errorMessage: "",
  tokens: null,
};

export const signup = createAsyncThunk(
  "user/signup",
  async ({ email, password, username }: RegisterProps) => {
    return await register({ email, password, username });
  }
);

export const signin = createAsyncThunk(
  "user/signin",
  async ({ email, password }: LoginProps) => {
    return await login({ email, password });
  }
);

export const token = createAsyncThunk(
  "user/token",
  async ({email, password} : GetTokenProps) => {
    return await getToken({email, password});
  }
);

// export const refresh = createAsyncThunk(
//   "user/refresh",
//   async (refresh: string) => {
//     return await refreshToken(refresh);
//   }
// )

const userSlice = createSlice({
  name: "user",
  initialState,
  reducers: {
    setUser: (state, action: PayloadAction<User>) => {
      state.userState = action.payload;
      saveUserToLocalStorage(state.userState);
    },
    setTokens: (state, action: PayloadAction<Tokens>) => {
      state.tokens = action.payload;
    },
    logout: (state) => {
      state.userState = null;
      state.isAuthState = false;
      removeUserFromLocalStorage();
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(signup.fulfilled, (state, action) => {
        state.userState = action.payload;
      })
      .addCase(signup.rejected, (state, action) => {
        state.errorMessage = "Ошибка: " + action.error.message;
      })
      .addCase(signin.fulfilled, (state, action) => {
        state.userState = action.payload;
      })
      .addCase(signin.rejected, (state, action) => {
        const error = action.error;
        console.log(error);
        state.errorMessage = "Ошибка: " + action.error.message;
      });
  },
});

export const { setUser, setTokens, logout } = userSlice.actions;
export const userReducer = userSlice.reducer;
