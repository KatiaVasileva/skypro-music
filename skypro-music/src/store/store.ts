import { combineReducers, configureStore } from "@reduxjs/toolkit"
import { trackReducer } from "./features/trackSlice"
import { playlistReducer } from "./features/playlistSlice"
import { TypedUseSelectorHook, useDispatch, useSelector, useStore } from "react-redux";

export const makeStore = () => {
    return configureStore({
        reducer: combineReducers({
            track: trackReducer,
            playlist: playlistReducer,
        }),
    });
};

export type AppStore = ReturnType<typeof makeStore>;
export type RootState = ReturnType<AppStore["getState"]>;
export type AppDispatch = AppStore["dispatch"];

export const useAppDispatch: () => AppDispatch = useDispatch;
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;
export const useAppStore: () => AppStore = useStore;
