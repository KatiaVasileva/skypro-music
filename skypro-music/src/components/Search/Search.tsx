"use client"

import { useAppDispatch, useAppSelector } from "@/store/store";
import Icon from "../Icon/Icon";
import styles from "./Search.module.css";
import { setSearchState } from "@/store/features/filterSlice";
import { ChangeEventHandler } from "react";

function Search() {
  const { searchState } = useAppSelector((state) => state.filter);
  const dispatch = useAppDispatch();

  const handleSearchChange: ChangeEventHandler<HTMLInputElement> = (event) => {
    dispatch(setSearchState(event.target.value));
  };

  return (
    <div className={styles.search}>
      <Icon iconClass={styles.searchSvg} name="icon-search" />
      <input
        className={styles.searchText}
        type="search"
        placeholder="Поиск"
        name="search"
        value={searchState}
        onChange={handleSearchChange}
      />
    </div>
  );
}

export default Search;
