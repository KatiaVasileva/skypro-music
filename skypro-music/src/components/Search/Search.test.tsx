import { fireEvent, render, screen } from "@testing-library/react";
import Search from "./Search";
import "@testing-library/jest-dom";
import { useAppDispatch, useAppSelector } from "@/store/store";
import { setSearchState } from "@/store/features/filterSlice";

jest.mock("../../store/store");

const mockUseAppDispatch = useAppDispatch as jest.Mock;
const mockUseAppSelector = useAppSelector as jest.Mock;
const mockDispatch = jest.fn();

describe("Search component", () => {
  beforeEach(() => {
    jest.clearAllMocks();
    mockUseAppDispatch.mockReturnValue(mockDispatch);
    mockUseAppSelector.mockImplementation((selector) =>
      selector({
        filter: {
          searchState: "",
        },
      })
    );
  });

  it("renders correctly", () => {
    const { container } = render(<Search />);
    expect(container).toMatchSnapshot();
    expect(screen.getByPlaceholderText("Поиск")).toBeInTheDocument();
  });

  it("should show current search value", () => {
    mockUseAppSelector.mockReturnValue({ searchState: "test" });
    render(<Search />);
    const input = screen.getByPlaceholderText("Поиск");
    expect(input).toHaveValue("test");
  });

  it("should dispatch setSearchState when input value changes", () => {
    mockUseAppSelector.mockReturnValue({ searchState: "" });
    render(<Search />);
    const input = screen.getByPlaceholderText("Поиск");
    fireEvent.change(input, { target: { value: "test" } });
    expect(mockDispatch).toHaveBeenCalledWith(setSearchState("test"));  
  });
});
