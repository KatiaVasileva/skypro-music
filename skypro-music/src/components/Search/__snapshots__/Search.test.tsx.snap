// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search component renders correctly 1`] = `
<div>
  <div
    class="search"
  >
    <div
      class=""
    >
      <svg
        class="searchSvg"
      >
        <use
          xlink:href="/img/icon/sprite.svg#icon-search"
        />
      </svg>
    </div>
    <input
      class="searchText"
      name="search"
      placeholder="Поиск"
      type="search"
      value=""
    />
  </div>
</div>
`;
